scripts/dts/python-devicetree/src/devicetree/edtlib.py

```diff
@@ -455,6 +474,126 @@ def _add_included_prop2specs(self, fname: str, contents: dict,
             if prop not in self.prop2specs:
                 self.prop2specs[prop] = spec
 
+    # Add child-binding definitions inherited from an included binding.
```

Comment:
Hrm, while this (and #65221) seems to work, it introduces quite some additional complexity to the code.

**TLDR;**

This and #65221 both replicate existing code smells and introduce new ones (structural duplication, broken functional encapsulation, feature envy of functions, premature optimization, the same internal properties being updated from several abstraction layers, therefore unclear abstraction levels, function names not clearly saying what the function does, ...).

I therefore have the strong feeling that these fixes are actually unnecessarily ugly workarounds for a deeper pre-existing design problem that we should solve instead.

On a side note: It seems to me that this specific change alters the inheritance approach for child bindings from override to merge. I'm quite new to edtlib, so I might be wrong here - if so, please ignore. But if I'm right then this is a breaking change. I think in that case we should ensure that what we assume to be "intended" is clarified with a broader audience (arch WG?) to ensure that we're not accidentally breaking something. And then we should document the breaking change in the migration README of the release.

**Detailed solution proposal**

I think the basic design problem is that we mix up two things:
- establishing binding file inheritance per compat
- attaching validation information derived from bindings to nodes

And then there are some minor design problems like:
- a lot of derived state that needs to be synchronized (we should try to get rid of some of those state duplications because they break the single-point-of-truth principle and make the code much harder to maintain)
- some of the entities are not properly normalized, which is the root cause why you cannot access the "right" path in the first place
- ...

Side note: While analyzing this problem I think I discovered another bug. It seems as if child bindings are only inherited from the parent rather than from all ancestors. But maybe I'm wrong on that. The code is really hard to follow. But this problem would automatically go away, too if we apply the design I propose.

**Fixing the binding model:**

I agree with you that it is not necessary to keep the full history of inheritance on file level but only the "result" of what you call "top level binding" (including inherited allow/block lists - exactly as you propose). IMO we should rather call those "compats" by normalization rules, though, because that's what they they seem to be keyed on.

I'd therefore call "top level bindings" `Compat` as a class from now on (inheriting from Binding) and introduce attributes and properties that are exclusive to "top level bindings" (most notably compat). Bindings are layers of validation rules (top level plus all child levels) and exclusively contain merged property specs and binding-level attributes (i.e. those also allowable for child bindings including merged allow/block lists at each level).

We also need to represent attributes of bindings (including compat) and property specs as conceptual entities in their own right, because they de-facto have at least name, value, overridable (bool), _and_ "last overriding source path" as their attributes, i.e. by normalization rules they have to be considered distinct relations. This allows us to simplify our merging process considerably as we can replace attribute instances when they are overridden. No need to track any derived state anywhere else. (This simple design fix will probably do away with most of the complexity of your fix.)

Note: All binding and property _attributes_ have to track their own path. We can implement non-overridable attributes by enforcing their path to be constant based on the "overridable" property via assertions on path updates. We may derive a (deprecated) class-level path property for backwards compat.

**Fixing the node model:**

Attaching bindings (compats) to nodes seems reasonable at first sight, because nodes specify compats. But once child bindings and allow/block lists enter the picture, they'll obviously add path dependent constraints to subnodes with distinct compats. Here lies the main design error.

I propose that we introduce a `Validation` class distinct from `Compat` and `Binding`. It represents a merged list of all allowed and blocked property specs applicable _per node_. These obviously come from different bindings (aka compats) and need to be filtered individually for each node.

We can establish those validations by...
* ...either applying binding levels from top to bottom to the whole device tree (via breadth first iteration of the tree) - I'd prefer that for simplicity.
* Alternatively we can continue to establish validations "on the fly" from nodes "upwards". But then we need to extend it to all anchestors and not only the parent. Such complexity should still be encapsulated in its own class so that validation clients don't have to deal with it. The only thing a validation client cares about is a list of property specs applicable to one node (including the corresponding meta-information for debugging and commenting results).

**Fixing the algorithm:**

Introducing this fixed class model requires us to (slightly) change the transformation process:

1. We first have to collect a deduplicated set of all compatibles from the full node tree iterating it in any order.
2. We iterate this list to establish all required top level bindings (which I'd call "Compatible" as a class from now on) including inheritance and key them on compats.
3. We iterate the device tree breadth first and apply the compat binding to the current node and the child layer bindings to all of the node's children recursively (iterated breadth first) until the child binding hierarchy of the binding is exhausted. This also allows us to allow a much cleaner validation logic with well-defined and easy-to-maintain overriding rules for child bindings vs. compat bindings.
4. We iterate the device tree depth first and filter the validations based on allow/block lists.

This algo will probably completely do away with the need to have a derived property2specs list. And I wouldn't be surprised if we found that other derived state can also be done away with in transit.

Reply:
Oh - of course I can help with the refactoring if you're afraid of such a structural fix. I'm onto some refactoring of edtlib anyway.